### PR TITLE
Fix #24466 and #24735: make HTTP/2 and HTTP/3 codecs send 400 for headers/tailers with underscores (if configured to reject)

### LINKED ISSUE #24466: H/2 codec resets streams rejected due to headers with underscores
If Envoy is configured to reject requests with underscores in header keys, the H/2 codec will reset the stream instead of sending 400 (like H/1).

Make this behavior consistent across all codecs, by responding with 400.

### PR DESCRIPTION
**Pull request by Winbobob**
Commit Message: Fix #24466 and #24735: make HTTP/2 and HTTP/3 codecs send 400 for headers/tailers with underscores (if configured to reject)
Additional Description: If Envoy is configured to reject requests with underscores in header/tailer keys, the HTTP/2 and HTTP/3 codecs will return 400 (like HTTP/1).
Risk Level: Low
Testing: Unit tests
Docs Changes:
Release Notes:
Platform Specific Features:
Fixes #Issue: #24466 and #24735

### CODE REVIEW

File: test/extensions/http/header_validators/envoy_default/http2_header_validator_test.cc [hunk 1]
```diff
@@ -438,15 +438,6 @@ TEST_F(Http2HeaderValidatorTest, ValidateResponseHeaderMapEmptyGenericName) {
                              UhvResponseCodeDetail::get().EmptyHeaderName);
 }
 
-TEST_F(Http2HeaderValidatorTest, ValidateRequestTrailersAllowUnderscoreHeadersByDefault) {
```
Comment (Winbobob) [inline test/extensions/http/header_validators/envoy_default/http2_header_validator_test.cc:441]: This test is identical to the https://github.com/envoyproxy/envoy/blob/a298895dd66a8df6ff834c5d646c38e0fc21e105/test/extensions/http/header_validators/envoy_default/http2_header_validator_test.cc#L729. So I removed this one.

**Comment by Winbobob** [comment]
Hi @yanavlasov, would you mind reviewing my source code changes?
I am thinking to make the source code changes correct before fixing test codes. Otherwise, tests may never pass.
First time to contribute hcm and http codec codebase, any suggestions are really appreciated!

**Comment by adisuissa** [comment]
@Winbobob please clarify the PR comment, and the PR subject (the fix part should be in the PR comment, as you have it, but the subject should describe what change is introduced by the PR).

**Comment by yanavlasov** [comment]
Thanks @Winbobob for taking on this work. However this PR does not fix the problem in issues #24466 and #24735. This requires changing how the check for underscores is done in H/2 and H/3 codecs. H/2 reference: https://github.com/envoyproxy/envoy/blob/40dddc9ee82e63ef06903efac822a08e031473b6/source/common/http/http2/codec_impl.cc#L1568

The change should affect the HeadersWithUnderscoresCauseRequestRejected test such that it does not need the `if (downstream_protocol_ == Http::CodecType::HTTP1)` statement.

/wait-any

**Comment by Winbobob** [comment]
> Thanks @Winbobob for taking on this work. However this PR does not fix the problem in issues #24466 and #24735. This requires changing how the check for underscores is done in H/2 and H/3 codecs. H/2 reference:
> 
> https://github.com/envoyproxy/envoy/blob/40dddc9ee82e63ef06903efac822a08e031473b6/source/common/http/http2/codec_impl.cc#L1568
> 
> The change should affect the HeadersWithUnderscoresCauseRequestRejected test such that it does not need the `if (downstream_protocol_ == Http::CodecType::HTTP1)` statement.
> 
> /wait-any

Hi @yanavlasov, I try to make h2 codec not reset the stream, but let `validateHeaders` on hcm to send local reply 400. Wonder does this approach sound correct to you?

**Comment by yanavlasov** [comment]
Very close, but not quite. We can not modify the existing behavior without UHV as it can affect existing operators. We also need a fully backward compatible mode when UHV is enabled. I think the best approach is to add a runtime flag that changes this behavior when UHV is enabled, and leave the non UHV behavior as is. You can see the example in #26741

/wait-any

**Pull request closed without merging**